airflow/jobs/triggerer_job_runner.py

```diff
@@ -314,95 +122,222 @@ def on_kill(self):
+
+        pretty_logs = False
+        if pretty_logs:
+            underlying_logger: WrappedLogger = structlog.WriteLogger(log_file.open("w", buffering=1))
+        else:
+            underlying_logger = structlog.BytesLogger(log_file.open("wb"))
+        logger = structlog.wrap_logger(underlying_logger, processors=processors).bind()
+        self.bound_logger = logger
+        return logger
+
+
+@attrs.define(kw_only=True)
+class TriggerRunnerSupervisor(WatchedSubprocess):
+    """
+    TriggerRunnerSupervisor is responsible for monitoring the subprocess and marshalling DB access.
+
+    This class (which runs in the main process) is responsible for querying the DB, sending RunTrigger
+    workload messages to the subprocess, and collecting results and updating them in the DB.
+    """
+
+    job: Job
+    capacity: int
+
+    health_check_threshold = conf.getint("triggerer", "triggerer_health_check_threshold")
+
+    runner: TriggerRunner | None = None
+    stop: bool = False
+
+    decoder: ClassVar[TypeAdapter[ToSyncProcess]] = TypeAdapter(ToSyncProcess)
+
+    # Maps trigger IDs that we think are running in the sub process
+    running_triggers: set[int] = attrs.field(factory=set, init=False)
+
+    logger_cache: dict[int, TriggerLoggingFactory] = attrs.field(factory=dict, init=False)
+
+    # A list of triggers that we have told the async process to cancel. We keep them here until we receive the
+    # FinishedTriggers message
+    cancelling_triggers: set[int] = attrs.field(factory=set, init=False)
+
+    # Outbound queue of events
+    events: deque[tuple[int, events.TriggerEvent]] = attrs.field(factory=deque, init=False)
+
+    # Outbound queue of failed triggers
+    failed_triggers: deque[tuple[int, list[str] | None]] = attrs.field(factory=deque, init=False)
+
+    def is_alive(self) -> bool:
+        # Set by `_service_subprocess` in the loop
+        return self._exit_code is None
+
+    @classmethod
+    def start(  # type: ignore[override]
+        cls,
+        *,
+        job: Job,
+        logger=None,
+        **kwargs,
+    ):
+        proc = super().start(id=job.id, job=job, target=cls.run_in_process, logger=logger, **kwargs)
+
+        msg = messages.StartTriggerer(requests_fd=proc._requests_fd)
+        proc._send(msg)
+        return proc
+
+    def _handle_request(self, msg: ToSyncProcess, log: FilteringBoundLogger) -> None:  # type: ignore[override]
+        if isinstance(msg, messages.TriggerStateChanges):
+            log.debug("State change from async process", state=msg)
+            if msg.events:
+                self.events.extend(msg.events)
+            if msg.failures:
+                self.failed_triggers.extend(msg.failures)
+            for id in msg.finished or ():
+                self.running_triggers.discard(id)
+                self.cancelling_triggers.discard(id)
+                # TODO: Close logger? Or is deleting it enough
+                self.logger_cache.pop(id, None)
```

Comment:
Should we close it here?

Reply:
If this is the last reference to the logger, then removing it from the dictionary will remove the last reference to it and then it'll get closed by the GC calling `__del__` on the object etc.

But it's probably more correct to close it, yes